### ISSUE TITLE
Can't delete failure.xhtml on Windows because FailureElement keeps a handle open to it

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5509]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| ivank reported|
|-|-|
|Trac ID|trac#5509|
|Type|defect|
|Created|2012-03-03 01:37:11Z|
|Branch|https://github.com/twisted/twisted/tree/xmlfile-closing-5509|

`twisted.web.util.FailureElement` keeps a handle open to `failure.xhtml`:

```
loader = XMLFile(FilePath(__file__).sibling('failure.xhtml').open())
```

On Windows, this makes it very hard to change the state of a Twisted checkout when I have a twisted.web program running.  If I install Twisted, this problem probably makes it impossible to fully install a new Twisted (again, while a program is running).

Process Explorer shows that my program does not keep handles open to other Twisted files, so this problem, despite being a Python open()/Windows issue, should probably be fixed.

Attachments:

* [5509-00-load-template-into-memory.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/554/5544a9c65fe1288084484177a840a6ac37117736/a7207e0de82cd6cf821a6057a3f81a5461e4b0f9.patch) (866 bytes) - added by ivank on 2012-03-03 01:57:19Z - don't leave a handle open to failure.xhtml
* [5509-01-XMLFile-FilePath-support.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/554/5544a9c65fe1288084484177a840a6ac37117736/62f83b69b18ebd976e9a160e68a7578f9d077e00.patch) (10830 bytes) - added by ivank on 2012-03-18 05:33:42Z - 
* [5509-02-XMLFile-FilePath-support.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/554/5544a9c65fe1288084484177a840a6ac37117736/c012dcb1cce201d5077cd053b113bea09c7bce6c.patch) (14390 bytes) - added by ivank on 2012-03-18 22:41:33Z - 
* [5509-03-XMLFile-FilePath-support.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/554/5544a9c65fe1288084484177a840a6ac37117736/d9474cb3f2920a3a0f429e554adec0cd130c386e.patch) (14570 bytes) - added by ivank on 2012-03-18 22:52:05Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5509 5509
type__defect defect
reporter__ivank ivank
priority__normal normal
milestone__ 
branch__branches_xmlfile_closing_5509 branches/xmlfile-closing-5509
branch_author__ivank ivank
status__closed closed
resolution__fixed fixed
component__core core
keywords__windows windows
time__1330738631000000 1330738631000000
changetime__1332176853000000 1332176853000000
version__None None
owner__exarkun exarkun
cc__ivank
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|ivank commented|
|-|-|

Also, unfortunately Python does not provide a way to open a file with the `FILE_SHARE_ALL` share mode, which would allow the open file to be deleted.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|ivank commented|
|-|-|

Sorry, by `FILE_SHARE_ALL` I meant `FILE_SHARE_DELETE | FILE_SHARE_READ | FILE_SHARE_WRITE`

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|ivank commented|
|-|-|

I added a .misc file because this problem hasn't made it into a release, and I doubt anyone else has noticed it in the month it's been there.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to ivank|
|-|-|

This fix makes me think that `XMLFile` should never be used, because it will always screw over Windows.

Perhaps instead we should change `XMLFile` so it's safe to use?  Perhaps `XMLFile` should immediately parse the template given to it and then close the file.

Or perhaps we should get rid of `XMLFile`, or replace it (or its `__init__`) with something that takes a `FilePath` instance instead of a `file` instance.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|ivank commented|
|-|-|

Here are my thoughts on three options:

1. Close the file immediately: closing whatever file object the user passes could be surprising and unwanted.

2. Make `XMLFile` take a `FilePath` instance that it can read and close: we'd have to support and deprecate the old behavior.  Not sure this is a good idea, given `XMLFile`'s marginal utility.

3. Deprecate `XMLFile` and change all uses of `XMLFile` to `XMLString`: this sounds okay to me.  It's curious how the docs have a bunch of incorrect uses of `XMLFile`, passing in a filename instead of a file object.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|ivank commented|
|-|-|

I just learned that `XMLFile` can also a string filename, though this is not documented in `XMLFile`'s docstring.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|ivank commented|
|-|-|

dreid convinced me to not get rid of `XMLFile`, because:

1. `XMLFile(FilePath(...))` is shorter than `XMLString(FilePath(...).getContent())`

2. `XMLFile` could later become capable of reloading templates.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|ivank commented|
|-|-|

I've made these changes:

1. `XMLFile` can now take a `FilePath`.

2. `XMLFile`'s support for file objects and filenames is deprecated.

3. I've updated the documentation to pass in `FilePath`s instead of filenames.

4. I've added a `__repr__` for `XMLFile`.

5. `FailureElement` now passes in a `FilePath`, solving the original problem.

6. `FailureElement` now uses `getModule`, per glyph's suggestion.

7. `FailureElement` now loads the template immediately, to prevent it from being out of sync with the Python code it's coupled with.

I did not document the `FailureElement` changes in NEWS files because they haven't made it into a release.

Please review.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|ivank removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst set owner to ivank|
|-|-|

(Review courtesy of JP, who is missing his credentials)

Thanks!

  1. Various docstrings need to be added (ie, for all methods)
  1. The `@param` doc in `XMLFile.__doc__` was wrong before, and is still wrong.  `@param` is for parameters.  This should have been `@ivar` before.  Maybe now it should be an `@ivar` on the class docstring and a `@param` in the `__init__` docstring (and they're not quite the same).
  1. Indentation on the `warnings.warn` call is weird; align args with the opening ( or put no args on the first line at all and indent by 4 spaces.
  1. I think I disagree with the idea of immediately loading the failure template in `twisted/web/util.py`.  If someone replaces the file from under us, do they have any grounds to expect sane behavior?  Part of the idea of accepting a path in `XMLFile` is to be able to _not_ read all of the templates at startup time.
  1. In `XMLFileReprTests`, the test method names can lose the _repr_ substring.  They test class already implies this.
  1. The `flushWarnings` use in `XMLLoaderTestsMixin` is a little fragile.  It will collect any warnings emitted during the test.  That may be none now, but if some new warning appears later on (eg, from the stdlib xml library) it will make the test fail.  I think that passing `offendingFunctions=[self.loaderFactory]` is a simple fix for this.  This both:
    1. Disregards warnings that point at some other call site as the cause
    1. Ensures that the warning emitted by `XMLFile` correctly points at the code constructing the `XMLFile` (ie, gets the `stacklevel` argument correct)

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|ivank removed owner|
|-|-|

Thanks for the review.

1. Okay, I think I've added all the missing docstrings.

2. Fixed

3. Fixed

4. Okay, undone

5. Fixed

6. Fixed, and confirmed that `offendingFunctions=` does something

I have also fixed some docstrings, including the `XMLString`, `TagLoader`, and `_flatsaxParse` docstrings.  I have also adjusted some vertical whitespace.

Please review (just the latest patch).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [33936]) Branching to 'xmlfile-closing-5509'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [33937]) Apply 5509-03-XMLFile-FilePath-support.patch 

refs [#5509](https://github.com/twisted/twisted/issues/5509)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

Thanks.  Test method docstrings don't need to say "Test that X."  They can just say "X."  Apart from that, I like the looks of this patch a lot.  If [ buildbot is similarly satisfied](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/xmlfile-closing-5509), I'll make those minor doc tweaks and land the change.  Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set status to `closed`|
|-|-|

(In [33940]) Merge xmlfile-closing-5509 

Author: ivank
Reviewer: exarkun
Fixes: [#5509](https://github.com/twisted/twisted/issues/5509)

Deprecate construction of `twisted.web.template.XMLFile` instances using a string
or a file object and add support for constructing them from a `FilePath` instance.
This also fixes a problem where, when a file object was used, that file object
would never be closed.
